docs/articles/Vector-Indexes.md

```diff
@@ -0,0 +1,401 @@
+
+### Hierarchical Navigable Small Worlds (HNSW)
+
+HNSW, a complex but popular, efficient vector indexing technique, builds layers of hierarchical proximity graphs structure like skip-lists, the lowest level with the highest granularity (every vector is a vertex), enabling precise, accurate navigation across short-range connections, and each higher layer vertex grouping (and connecting to) more and more lower level data points together (representing longer and longer connections between vertices). HNSW queries are rapidly ANN-matched to closest vectors in the highest layer, then again with slightly more accuracy in the next layer, and so on until search results are surfaced with the greatest accuracy from the lowest layer.
+
+Here's a table summarizing these common vector indexing approaches in terms of speed, accuracy, database size, and memory usage.
+
+| Indexing Method | Speed&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Accuracy&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Database Size | Memory Usage&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; |
+| :--- | :--- | :--- | :--- | :--- |
+| Flat | slow : query compared with each element | high : exact matches | small to medium | high : vectors stored as is in memory |
+| LSH | fast : hashing queries into buckets for quicker retrieval | moderate : approximate matches | large | low : stores hashes instead of vectors |
+| IVF_FLAT | moderate : queries only relevant clusters, but varies with cluster size | high : performs exact matches in the cluster | small to medium | high : vectors stored as is in memory |
+| IVF_PQ | fast : clustering and compressing vectors into quantized codes for efficient retrieval | moderate : dependent on the quantization level | large | low : compressed representation |
+| IVF_SQ | fast : clustering and compressing vectors into quantized codes for efficient retrieval | moderate : approximate matches in quantized space | large | low : compressed representation |
+| DiskANN | fast : uses high-speed SSDs and a special Graph Data Structure | high (95% reported in original paper) | very large | moderate : storing graphs on disk |
+| SPANN | fast : uses disk-memory hybrid structure | high (90% reported in original paper) | very large | moderate : uses a memory-disk hybrid structure |
+| HNSW | fast : hierarchical graph structure | moderate : progressive refinement of approximate results, improved with parameter optimization | large | moderate : graph structure in memory |
```

Comment:
@haziqa5122
I think this table needs numbers to work. Can you take a close look at these units and let me know if this is roughly accurate:

Indexing Method	Speed	Accuracy	Database Size	Memory Usage
Flat	slow: ~10 ms/query	high: 100%	small to medium: up to 10M vectors	high: ~1 GB for 1M vectors
LSH	fast: ~1 ms/query	moderate: ~80%	large: up to 100M vectors	low: ~100 MB for 1M vectors
IVF_FLAT	moderate: ~5 ms/query	high: 95%	small to medium: up to 10M vectors	high: ~1 GB for 1M vectors
IVF_PQ	fast: ~2 ms/query	moderate: ~85%	large: up to 100M vectors	low: ~100 MB for 1M vectors
IVF_SQ	fast: ~2 ms/query	moderate: ~85%	large: up to 100M vectors	low: ~100 MB for 1M vectors
DiskANN	fast: ~1 ms/query	high: 95%	very large: up to 1B vectors	moderate: ~500 MB for 1M vectors
SPANN	fast: ~1 ms/query	high: 90%	very large: up to 1B vectors	moderate: ~500 MB for 1M vectors
HNSW	fast: ~1 ms/query	moderate: ~90%	large: up to 100M vectors	moderate: ~200 MB for 1M vectors